android/src/com/unciv/app/CustomSaveLocationHelperAndroid.kt

```diff
@@ -0,0 +1,115 @@
+// https://developer.android.com/guide/topics/providers/document-provider
+@RequiresApi(Build.VERSION_CODES.KITKAT)
+class CustomSaveLocationHelperAndroid(private val activity: Activity) : CustomSaveLocationHelper {
+    // This looks a little scary but it's really not so bad. Whenever a load or save operation is
+    // attempted, the game automatically autosaves as well (but on a separate thread), so we end up
+    // with a race condition when trying to handle both operations in parallel. In order to work
+    // around that, the callbacks are given an arbitrary index beginning at 100 and incrementing
+    // each time, and this index is used as the requestCode for the call to startActivityForResult()
+    // so that we can map it back to the corresponding callback when onActivityResult is called
+    @GuardedBy("this")
+    @Volatile
+    private var callbackIndex = 100
+    @GuardedBy("this")
+    private val callbacks = ArrayList<IndexedCallback>()
+
+    override fun saveGame(gameInfo: GameInfo, gameName: String, forcePrompt: Boolean, block: (() -> Unit)?) {
+        val callbackIndex = synchronized(this) {
+            val index = callbackIndex++
+            callbacks.add(IndexedCallback(
+                    index,
+                    { uri ->
+                        uri?.let {
+                            saveGame(gameInfo, it)
+                        }
+                        block?.invoke()
+                    }
+            ))
+            index
+        }
+        gameInfo.customSaveLocation?.let { customSaveLocation ->
+            if (!forcePrompt) {
+                handleIntentData(callbackIndex, Uri.parse(customSaveLocation))
+                return
+            }
+        }
+
+        Intent(Intent.ACTION_CREATE_DOCUMENT).apply {
+            type = "application/json"
+            putExtra(Intent.EXTRA_TITLE, gameName)
+            activity.startActivityForResult(this, callbackIndex)
+        }
+    }
+
+
+    // This will be called on the main thread
+    fun handleIntentData(requestCode: Int, uri: Uri?) {
+        val callback = synchronized(this) {
+            val index = callbacks.indexOfFirst { it.index == requestCode }
+            if (index == -1) return
+            callbacks.removeAt(index)
+        }
+        callback.thread.run {
+            callback.callback(uri)
+        }
+    }
+
+    private fun saveGame(gameInfo: GameInfo, uri: Uri) {
+        gameInfo.customSaveLocation = uri.toString()
+        activity.contentResolver.openOutputStream(uri, "rwt")
+                ?.writer()
+                ?.use {
+                    it.write(json().toJson(gameInfo))
+                }
+    }
+
+    override fun loadGame(block: (GameInfo) -> Unit) {
+        val callbackIndex = synchronized(this) {
+            val index = callbackIndex++
+            callbacks.add(IndexedCallback(
+                    index,
+                    { uri ->
+                        uri?.let {
```

Comment:
Same points - nullability check should be with if()s and not with ?. or .let{}s - https://github.com/yairm210/Unciv/wiki/Coding-standards